src/Adapter/Shipment/DeliveryOptionsProvider.php

```diff
@@ -0,0 +1,183 @@
+        PriceFormatter $priceFormatter
+    ) {
+        $this->context = $context;
+        $this->objectPresenter = $objectPresenter;
+        $this->translator = $translator;
+        $this->priceFormatter = $priceFormatter;
+    }
+
+    private function isFreeShipping(array $deliveryOption)
+    {
+        if ($deliveryOption['is_free']) {
+            return true;
+        }
+
+        $cart = $this->context->cart;
+
+        foreach ($cart->getCartRules() as $rule) {
+            if ($rule['free_shipping'] && !$rule['carrier_restriction']) {
+                return true;
+                break;
+            }
+        }
+
+        return false;
+    }
+
+    public function getSelectedDeliveryOption()
+    {
+        return current($this->context->cart->getDeliveryOption(null, false, false));
+    }
+
+    public function getDeliveryOptions()
+    {
+        $deliveryOptions = $this->context->cart->getDeliveryOptionList();
+        $currentAddressDeliveryOptions = $deliveryOptions[$this->context->cart->id_address_delivery];
+
+        if (empty($deliveryOptions[$this->context->cart->id_address_delivery])) {
+            return [];
+        }
+
+        $formattedDeliveryOptions = [];
+
+        foreach ($currentAddressDeliveryOptions as $deliveryOptionId => $deliveryOption) {
+            $formattedDeliveryOption = end($deliveryOption['carrier_list']);
+            $formattedDeliveryOption = array_merge($formattedDeliveryOption, $this->objectPresenter->present($formattedDeliveryOption['instance']));
+            unset($formattedDeliveryOption['instance']);
+
+            $carriersDetails = $this->getCarriersDetails($deliveryOption);
+            $formattedDeliveryOption = array_merge($formattedDeliveryOption, $carriersDetails);
+
+            $formattedDeliveryOption['price'] = $this->getPriceToDisplay($deliveryOption);
+            $formattedDeliveryOption['label'] = $formattedDeliveryOption['price'];
+
+            $formattedDeliveryOptions[$deliveryOptionId] = $formattedDeliveryOption;
+        }
+
+        return $formattedDeliveryOptions;
+    }
+
+    private function getCarriersDetails($deliveryOption)
+    {
+        $carriers = $deliveryOption['carrier_list'];
+
+        if (count($carriers) === 1) {
+            return [
+                'id' => $carriers[0]['instance']->id,
+                'label' => $carriers[0]['label'],
+                'name' => $carriers[0]['instance']->name,
+                'delay' => $carriers[0]['instance']->delay[$this->context->language->id],
+            ];
+        }
+
+        $names = [];
+        $delays = [];
+        $extraContent = '';
+
+        foreach ($carriers as $carrier) {
+            $names[] = $carrier['instance']->name;
+            $delays[] = $carrier['instance']->delay[$this->context->language->id];
+            $extraContent .= Hook::exec('displayCarrierExtraContent', ['carrier' => $carrier['instance']], Module::getModuleIdByName($carrier['instance']->id));
```

Comment:
One small comment here, the logic is a bit different in the original code.

```
// If carrier related to a module, check for additionnal data to display
                            $carrier['extraContent'] = '';
                            if ($carrier['is_module']) {
                                if ($moduleId = Module::getModuleIdByName($carrier['external_module_name'])) {
                                    // Hook called only for the module concerned
                                    $carrier['extraContent'] = Hook::exec('displayCarrierExtraContent', ['carrier' => $carrier], $moduleId);
                                }
                            }
```

Reply:
Indeed, it could be nice. Added 👍🏻